### PR TITLE
LIBTREATDB-56 Replace Devise with Custom Session Management - FEATURE BRANCH

### PR DESCRIPTION
**Pull request by Janell-Huyck**
https://ucdts.atlassian.net/browse/LIBTREATDB-56

## There are instructions for migrations and starting the app at the bottom of this description.

---

## Do not follow these instructions for testing this PR.  See the bottom of this PR description for instructions.

If you've been reviewing other PR's, you have likely had to run `rails db:migrate` for this one.  Now, if you're going back to that branch, your database is in a different state and you will be unable to run or test the program.  You will need to reset your databases.  Here is a simple way to do what is needed.  From your terminal, navigate to the branch you are reviewing, and then type (or copy/paste):

```
RAILS_ENV=development rails db:drop db:create db:schema:load db:seed
RAILS_ENV=test rails db:drop db:create db:schema:load db:seed
```

If you want to check things in the production environment, you would use the same line of commands for production:

```
RAILS_ENV=production rails db:drop db:create db:schema:load db:seed
```
--- 

### Gem changes:
- Removes Devise
- Adds Bcrypt for handling password verification

### Routes changes
- remove Devise routes
- add sessions routes
- add admin/user_controller routes for admins to manage users
- limit "user" routes to functions normal users can do

### Controllers changes
- admin/users_controller
  - New controller, accessible only to admin accounts
  - used to create or edit any user account, (email, display name, password, active-status, role)
  - cannot delete an account
- users_controller
  - Updated, accessible to any user
  - update own display name (only)
  - cannot update own email , password, or username, in line with how Shibboleth will work
  - removed Devise-specific functionality now handled by admin/users_controller
- registration_controller
  - removed since this was a Devise-only controller
- sessions_controller
  - finds users
  - verifies active status and password checks
  - assigns session id's to the user ids
  - destroys sessions
- front_controller (shows the home page & is the space to login)
  - set it so we aren't validating login status when visiting this page
- concern: authentication
  - included in application_controller
  - doesn't allow the user to visit any pages if not logged in and active

### Helpers changes
- new helper: navigation_helper
  - shows one of two navigation bars based on user's login status
  - determines what links will be shown to the user based on their CanCan permissions

### Mailers
- Removes application_mailer.rb. This was a Devise mailer, and we do not anticipate needing to email users for anything in the future since password issues will be handled through Shibboleth.

### Models
- CanCan Ability (model):
  - New User Permissions: For both 'standard' and 'read_only' roles, new permissions were added. Users can now:
    - Read and update their own User record
    - They are also explicitly forbidden from accessing the index action on the User model
- Users Model
  - Adds new validations:
    - email: 
      - must end in @uc.edu or @ucmail.uc.edu
      - must be unique
      - must be present
    - display name: 
      - can only contain alphanumeric characters and space, period, dash, or underscore. 
      - must be present
      - must be unique
      - max-length 30 characters.
    - username: 
      - can only contain alphanumeric characters, period, dash, or underscore (no space). 
      - must be unique 
      - must be present.
    - role: must be present, must be included in a frozen list
  - "activate_new_account" and "default_role" logic now part of admin_users model & controller
  - "active_for_authentication" logic now part of authentication concern

### Views changes
- admin user functionalities
  - only viewable by admin users
  - new: "_form" for creating or editing a new user
    - (This will need to be updated to not include password once Shibboleth is working. [LIBTREATDB-83](https://ucdts.atlassian.net/browse/LIBTREATDB-83))
  - "edit" wrapper for "_form" for editing user
  - "new" wrapper for "_form" for creating user
    - (This will need to be updated to not include password once Shibboleth is working. [LIBTREATDB-83](https://ucdts.atlassian.net/browse/LIBTREATDB-83))
  - "index" - removed unnecessary permission check, updated link to edit & create
- removed Devise views
- navbar: updated links
- sessions: new login page since Devise view isn't being used
- regular users: updated to only pertain to users updating their own display name and viewing email and role

### Config changes
- config/environments/test.rb: remove reference to Devise
- Devise initializers: removed

### Routes
- reconfigued users and admin/users routes to align with new functionality
- added sessions routes
- removed Devise routes

### Migrations
- remove devise-specific user fields
- add column "username" to users and fills in current data with default username
- adds new column constraints to users email, username, display name, role, account_active, and password_digest

### Seeds
- updates the seeded emails to all be @uc.edu

### Tests
The changes to testing are significant since Devise was used in all tests that required logging in or authentication. A few highlights:

- New helpers
  - controller, request, view, system authentication helpers stub logging in and authorization for their respective tests
  - spec/support/shoulda_matchers.rb - initializing shoulda matches used in abbreviated_treatment_report_spec.rb
- Changes to controller tests
  - Added sessions_controller tests
  - Remove Devise helpers and replace with custom helpers (above)
  - New tests for admin-specific user functionality
  - spec/controllers/users_controller_spec.rb - modified to only test functionality available to non-admins since admin functionality is now in spec/controllers/admin_users_controller_spec.rb
- New view tests:
  - Split navigation view tests into logged_in and logged_out tests
- Changes to factories
  - abbreviated_treatment_report: if not already associated with a conservation record, will create one and associate with it.
  - reports: it used to reference a non-existent file. Created a file and placed in "fixtures"
  - users:  
    - Introduces a transient attribute n using a custom sequence user_sequence. This stores the sequence value and is used for generating the username, email, and display_name fields.
    - New Fields:
      - password_confirmation: Added to match the password for cases where password confirmation is required.
      - username: Added and generated using the transient attribute n and timestamp, ensuring a unique username for each user.
    - The email domain was updated from example.com to uc.edu.
- Remove Devise-specific tests
  - spec/mailers/application_mailer_spec.rb
  - spec/mailers/devise_mailer_spec.rb
- Model tests
  - spec/models/ability_spec.rb - refactored to allow easier maintenance

### Spec helper
- Removed Devise setup

### Before running rails db:migrate:
<h2>IMPORTANT</h2>
<h3>Prepare your database</h3>
You probably have some seeded or old data that will break the app and needs to be fixed BEFORE you run migrations.  The user emails now MUST end in @uc.edu or @ucmail.uc.edu.  You can update this via `rails console` or from within the app.  The easiest way is probably through the app, like this: 

- Load up the qa branch and log in as an admin.  (chuck@chuck.codes, password 'notapass' is a default user)
- Navigate to the "Users" menu and for each of your users update the email domain to be @uc.edu or @ucmail.uc.edu. 

You will want to check that running the migrations doesn't destroy the data already in the user's database.
- Navigate to the Conservation Records page and change something that you can recognize as your own change - it can be adding a new record, modifying a record, and/or deleting a record.
- Add a new user or modify a user, remembering to keep the email domain as @uc.edu or @ucmail.uc.edu.

<h3>Run the migrations</h3>

- Run the migrations for the development and test environment
  - RAILS_ENV='development' rails db:migrate
  - RAILS_ENV='test' rails db:migrate 

Note: do NOT drop and re-create the databases. You are testing that the data remains intact with the migration.

<h3>Update dependencies and pages</h3>

- `bundle install`
- `yarn install`
- `yarn build`
- Start up the server with `rails server`.

<h3>Quick check</h3>

- You should be able to log in with the updated @uc.edu mail address.
- Your extra data should remain present and unchanged.

### CODE REVIEW

**Comment by hortongn** [comment]
Suggested strategy for reviewing this PR:
1. A couple people get this running in their local environments and do an end-to-end manual test of the app.  Verify all behavior as best as possible.  Verify existing data is unaltered.
2. After any needed changes are made, do a code review on this.
3. Deploy to libappstest and do another run-through of the app.

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    if user_authenticated?(user)
+      login_user(user)
+    else
+      handle_invalid_credentials
+    end
+  end
+
+  def destroy
+    session[:user_id] = nil
+    redirect_to root_path, notice: 'Logged out successfully'
```
Comment (hortongn) [inline app/controllers/sessions_controller.rb:19]: I see you have a controller test for #destroy, but Coveralls is saying that the two lines in your destroy method are not covered for some reason.

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  skip_before_action :authenticate_user!, only: %i[new create]
+  skip_before_action :check_user_active, only: %i[new create]
+  def new; end
+
+  def create
+    user = find_user
+    if user_authenticated?(user)
+      login_user(user)
+    else
+      handle_invalid_credentials
```
Comment (hortongn) [inline app/controllers/sessions_controller.rb:13]: `handle_invalid_credentials` is untested

File: app/controllers/admin/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    def update
+      if @user.update(user_params)
+        redirect_to admin_users_path, notice: 'Profile updated successfully.'
+      else
+        render :edit
+      end
+    end
+
+    def destroy
+      redirect_to admin_users_path, alert: 'User deletion is not permitted.'
```
Comment (hortongn) [inline app/controllers/admin/users_controller.rb:42]: #destroy doesn't have a test.

File: app/controllers/admin/users_controller.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        flash.now[:alert] = 'There was a problem creating the user. Please check the errors below.'
+        render :new
+      end
+    end
+
+    def update
+      if @user.update(user_params)
+        redirect_to admin_users_path, notice: 'Profile updated successfully.'
+      else
+        render :edit
```
Comment (hortongn) [inline app/controllers/admin/users_controller.rb:37]: This else condition is untested.

File: app/controllers/concerns/.keep
Comment (hortongn) [inline app/controllers/concerns/.keep:1]: Do we need this .keep file?

**Review by hortongn — CHANGES_REQUESTED** [review]
@Janell-Huyck I added some inline comments about test coverage, but other than that I everything else looks good.  I found no problems running the app.

**Comment by scherztc** [comment]
I ran through the Manuel Test Script for each role and the only confusing thing for me was the User Edit screen as Admin.

I was confused on whether I needed to type in the Admin password to make the changes or whether or not I was changing the users password.   This confusion needs to be cleared up.

Also, How does a user change their own password?

Screenshot 2024-09-09 at 1 10 50 PM

**Review by scherztc — APPROVED** [review]

**Pull request merged**